[Mailer] Add support for rate limited mailer transports

fritzmg opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | 
| License       | MIT

Assume you have to use multiple SMTP servers in your application in order to be able to send emails on behalf of specific email accounts on a variety of hosting providers. Some of these hosting providers may impose limits on how many emails you can send per timeframe - especially in shared hosting environments. For example on Hetzner shared hostings the limit is [500 emails per hour](https://www.hetzner.com/legal/webhosting/#:~:text=The%20limit%20of%20500%20emails%20sent%20per%20hour%20must%20not%20be%20exceeded.), on webgo [50 emails per hour](https://webgo.de/faq/wie-hoch-ist-das-mail-limit-versand-limitierung-im-shared-hosting#:~:text=200%20Mails%20pro%204%20Stunden%20je%20Postfach). Hetzner does not force this limit - but they may disable your account completely, if you break this rule. The same applies to services like sendgrid etc., where you will have certain limits.

Now, if you use `symfony/messenger` in order to send your emails asynchronously, you are able to rate limit the messenger transport:

```yaml
framework:
    messenger:
        transports:
            mailer:
                dsn: 'doctrine://default?queue_name=mailer'
                rate_limiter: mailer
        routing:
            'Symfony\Component\Mailer\Messenger\SendEmailMessage': mailer
    rate_limiter:
        mailer:
            policy: fixed_window
            limit: 500
            interval: 60 minutes
```

However, this will only rate limit the email queue globally - and not per SMTP server. So if you have to use multiple SMTP servers, where some use different rate limits and others do not need to be rate limited at all, this is not ideal.

Thus this PR introduces support for rate limited mailer transports. The changes are fairly straight forward:

* For each mailer transport you can now also optionally define a `rate_limiter`, just as with messenger transports.
* This rate limiter will then be set by the `Transport` factory, if applicable.
* The actual rate limiting is done in the `AbstractTransport::send()` method - using `ensureAccepted()`, so that a `RateLimitExceededException` will be thrown.
* This means that - if you do not use a messenger queue for sending emails, your application is responsible for handling this exception, when you defined a rate limiter on a mailer transport.
* If you do use a messenger queue then the email message will be put back on the queue, with `available_at` set to the respective time in the future according to the rate limit.

```yaml
framework:
    mailer:
        transports:
            hetzner: 
                dsn: smtp://foo:bar@mail.your-server.de:587
                rate_limiter: hetzner
            webgo:
                dsn: smtp://lorem:ipsum@v123.goserver.host:587
                rate_limiter: webgo
            not_rate_limited: smtp://dolor:sit@example.com:587

    rate_limiter:
        # In actuality you will want to set a lower rate limit, depending on whether
        # other applications will also send via the same credentials.
        hetzner:
            policy: fixed_window
            limit: 500
            interval: 60 minutes
        webgo:
            policy: fixed_window
            limit: 50
            interval: 60 minutes

    messenger:
        transports:
            mailer: 
                dsn: 'doctrine://default?queue_name=mailer'
                failure_transport: mailer_failed
            mailer_failed: 'doctrine://default?queue_name=mailer_failed'
        routing:
            'Symfony\Component\Mailer\Messenger\SendEmailMessage': mailer
```

Tests are still missing, as I wanted to hear comments first :)

fritzmg committed:
First draft

fritzmg committed:
Fix retryDelay

OskarStark commented on src/Symfony/Component/Mailer/Messenger/MessageHandler.php:
```diff
@@ -26,6 +28,12 @@ public function __construct(
+        } catch (RateLimitExceededException $e) {
+            $retryDelay = ($e->getRetryAfter()->getTimestamp() - time()) * 1000;
+
+            throw new RecoverableMessageHandlingException('Rate limit for mailer transport exceeded.', previous: $e, retryDelay: $retryDelay);
```
we should avoid using named arguments here

fritzmg committed:
do not use named arguments

Spomky commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -2781,7 +2781,26 @@ private function registerMailerConfiguration(array $config, ContainerBuilder $co
+
+        foreach ($transports as $name => $transport) {
+            if ($transport['rate_limiter']) {
+                if (!interface_exists(LimiterInterface::class)) {
```
This `if` could be moved outside the `for` loop.
The verification can be done just after if `$transportRateLimiterReferences` is not empty and the interface is missing.

fritzmg replied:
Yes - however, the same would apply to this existing code: 

https://github.com/symfony/symfony/blob/3ec30b693fee9a37cdae6718259395eba2e8b25e/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L2458-L2460

Do you want me to change it in the unrelated section as well?

fritzmg replied:
Nevertheless, changed in 1f2714ee99b15a12107e2cdca33ae22de197f5b5

Spomky commented on src/Symfony/Component/Mailer/Transport/AbstractTransport.php:
```diff
@@ -24,13 +24,15 @@
 abstract class AbstractTransport implements TransportInterface
 {
     private LoggerInterface $logger;
+    private ?RateLimiterFactoryInterface $rateLimiterFactory = null;
```
New argument should be placed after to avoid BC breaks.

fritzmg replied:
Hm, how is this related to BC? This is just a private member variable - and not a constructor variable with constructor property promotion.

fritzmg replied:
Nevertheless changed in 8477ea1eb2

Spomky commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -2781,7 +2781,26 @@ private function registerMailerConfiguration(array $config, ContainerBuilder $co
+        $transportRateLimiterReferences = [];
+
+        foreach ($transports as $name => $transport) {
+            if ($transport['rate_limiter']) {
```
`isset` or `array_key_exists`?

fritzmg replied:
Not necessary as the key will always exist due to the processed configuration. See also: 

https://github.com/symfony/symfony/blob/3ec30b693fee9a37cdae6718259395eba2e8b25e/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L2457

Spomky reviewed (commented):
Hi,

Many thanks for the proposal. THis is a nice and useful idea.
I reviewed this PR and it is a good start except the missing tests and a few things I spotted.

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -2781,7 +2781,26 @@ private function registerMailerConfiguration(array $config, ContainerBuilder $co
+
+        foreach ($transports as $name => $transport) {
+            if ($transport['rate_limiter']) {
+                if (!interface_exists(LimiterInterface::class)) {
```
shouldn't this rather check `isConfigEnabled()` for the `rate_limiter` config ? the fact that the component is installed does not guarantee that the config is enabled for it (smart defaults are based on the availability of the component, but projects can still configure things explicitly)

fritzmg replied:
I am using the same logic as here: 

https://github.com/symfony/symfony/blob/3ec30b693fee9a37cdae6718259395eba2e8b25e/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L2421-L2464

Should this not employ the same logic as the `rate_limiter` config for `messenger` for consistency?

fritzmg replied:
Nevertheless, I have adjusted the implementation so that is is akin to

https://github.com/symfony/symfony/blob/6d2d7fa526a42b7a190628c53f4efb7a99ac431a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L403-L411

stof commented on src/Symfony/Component/Mailer/Transport/AbstractTransport.php:
```diff
@@ -103,6 +111,13 @@ public function send(RawMessage $message, ?Envelope $envelope = null): ?SentMess
+    {
+        $this->rateLimiterFactory = $rateLimiterFactory;
+
+        return $this;
```
I don't think we should return `$this` here. It makes the API less clear whether this is a method mutating the instance or returning a new one. Our setters generally don't return `$this` in Symfony.

fritzmg replied:
The setters in `Symfony\Component\Mailer\Transport\AbstractTransport` all return `$this` - so I think we should also return `$this` here, otherwise it would be inconsistent, would it not?

fritzmg replied:
Keeping as is, as `AbstractTransport` always returns `$this`.

fritzmg committed:
Move LimiterInterface out of loop

fritzmg committed:
move new member variable downwards

fritzmg committed:
use readConfigEnabled

fritzmg committed:
fix typo

fritzmg committed:
use isInitializedConfigEnabled

fritzmg committed:
Update tests

fritzmg committed:
Remove LimiterInterface

fabpot commented on src/Symfony/Component/Mailer/CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@ CHANGELOG
  * Add DSN param `source_ip` to allow binding to a (specific) IPv4 or IPv6 address.
  * Add DSN param `require_tls` to enforce use of TLS/STARTTLS
  * Add `DkimSignedMessageListener`, `SmimeEncryptedMessageListener`, and `SmimeSignedMessageListener`
+ * Add support for rate limited transports
```
Should be moved to the 7.4 section.

fabpot commented on .phpunit-12.1-0.md5:
To be removed

fritzmg committed:
move CHANGELOG entry

fritzmg committed:
Remove accidental PHPunit hash commit

fritzmg committed:
Add Transport test

fritzmg committed:
update CHANGELOG

fritzmg committed:
update composer.json

fritzmg committed:
code style

fabpot commented on src/Symfony/Component/Mailer/Transport.php:
```diff
@@ -93,26 +97,37 @@ public static function fromDsns(#[\SensitiveParameter] array $dsns, ?EventDispat
         foreach ($dsns as $name => $dsn) {
-            $transports[$name] = $this->fromString($dsn);
+            try {
+                $rateLimiter = $this->rateLimiterLocator?->get($name);
```
Do we need some sort of a prefix between the name of the transport and the rate limiter?
Like if the name of the transport is `default`, we would look for the `mailer.default` rate limiter?

fritzmg replied:
The logic is the same as for the Messenger rate limiters, no prefix is used there.

fritzmg committed:
[Mailer] Add a rate_limiter option to mailer transports

Providers cap how many messages an account may send per timeframe. Rate
limiting a Messenger transport only covers messages that are sent
asynchronously, and it cannot give two SMTP transports different caps, so a
transport now accepts a rate limiter of its own and fails fast with a
RateLimitExceededException when the limit is exceeded. When mails are sent
through Messenger, the handler turns that failure into a delayed retry so
the message is delivered once the limit resets.

nicolas-grekas commented:
Thanks for your patience @fritzmg. Rebased on 8.2, squashed and pushed.

**@fabpot, on your question about prefixing the rate limiter with the transport name:** no prefix, and the branch already had it right. It resolves `rate_limiter: foo_limiter` to the `limiter.foo_limiter` service, which is exactly what Messenger does for a transport's rate limiter. Same line, same shape:

```php
// FrameworkExtension, messenger transports (existing)
$transportRateLimiterReferences[$name] = new Reference('limiter.'.$transport['rate_limiter']);
// FrameworkExtension, mailer transports (this PR)
$transportRateLimiterReferences[$name] = new Reference('limiter.'.$transport['rate_limiter']);
```

`limiter.` is the service id every `framework.rate_limiter.*` entry is registered under, not a new namespace. Prefixing with the transport name would also stop a limiter being shared, which is what you want when a provider's cap applies to the whole account rather than one transport.

Your other two points are settled: the stray `.phpunit-12.1-0.md5` was already gone, and the CHANGELOG entry now sits under `8.2` in both Mailer and FrameworkBundle.

**What the rebase needed.** XML config was removed from FrameworkBundle in 8.2, so the hunk touching `Resources/config/schema/symfony-1.0.xsd` could not apply and the XML fixture went with it; the PR no longer touches either. The php and yml fixtures still declared `annotations: false`, which no longer exists under `framework`, so both `PhpFrameworkExtensionTest::testMailerRateLimiter` and the Yaml one errored on an unrecognized option. There was also a stray comma in the yml fixture's `interval: 5 seconds,`.

In `TransportTest::testSetsRateLimiter`, `createMock()` on `RateLimiterFactoryInterface` had no expectations, which is a PHPUnit notice now, so it is a stub; the `AbstractTransport` mock does set expectations on one of the two DSNs, so it stays a mock under `#[AllowMockObjectsWithoutExpectations]`.

Mailer is green at 956 tests and the FrameworkBundle DI suite at 582. php-cs-fixer is clean.

nicolas-grekas commented:
@fritzmg heads up on a behaviour change I pushed, please confirm you are happy with it before this gets merged.

The transport consumed the limiter and failed when it was over budget:

```php
$rateLimiter?->consume(1)->ensureAccepted();   // throws RateLimitExceededException
```

It now paces instead:

```php
$rateLimiter?->reserve()->wait();
```

My reasoning is that both precedents wait rather than throw:

* `Worker::rateLimit()` in Messenger consumes, and when the token is refused it dispatches `WorkerRateLimitedEvent` and calls `$rateLimiter->reserve()->wait()`;
* `AbstractTransport` itself, the very class this touches, already throttles by sleeping in `setMaxPerSecond()`: `$sleep = (1 / $this->rate) - (microtime(true) - $this->lastSent); usleep(...)`.

For the case in your description, a host that allows 500 mails an hour, throwing means the 501st message fails rather than being sent a little later, and a synchronous send loses it unless the caller catches the exception. "Rate limited transport" reads as pacing to me.

The trade is that a send can now block. That is already true of `setMaxPerSecond()`, and the option is opt-in, but it is a real difference: with a low limit and a synchronous send you can sit in a long `usleep`. If you would rather keep failing fast, or want `reserve(1, $maxTime)` so it waits up to a bound and throws beyond it, say so and I will change it back. No test asserted either behaviour, so nothing forced the choice.

Also in this push: `setRateLimiterFactory()` gained the summary line and `@return $this` that its neighbour `setMaxPerSecond()` carries, and the config option's `info()` now says it paces the messages rather than "to use when sending".

Mailer is green at 956 tests, FrameworkBundle DI at 582, php-cs-fixer clean.

fritzmg commented:
My idea was rather to then take advantage of the retry policy of the messenger transport in such a case - or the `$retryDelay` of the `RecoverableMessageHandlingException` for messenger transports that support it.

This would now block the whole messenger queue, would it not? Potentially for a long time if you have a sudden influx of a lot of email messages (say from a sign-up form or a newsletter). And in order to avoid that you would need to define a separate messenger transport anyway - which kind of defeats the purpose of this PR in my opinion. At least that's my first thought.

fritzmg commented:
It needs to stay

```php
$rateLimiter?->consume(1)->ensureAccepted();
```

Otherwise having 

```php
try {
    return $this->transport->send($message->getMessage(), $message->getEnvelope());
} catch (RateLimitExceededException $e) {
    $retryDelay = ($e->getRetryAfter()->getTimestamp() - time()) * 1000;

    throw new RecoverableMessageHandlingException('Rate limit for mailer transport exceeded.', 0, $e, $retryDelay);
}
```

In the `MessageHandler` makes no sense anymore.

nicolas-grekas commented:
Agreed, reverted to `consume(1)->ensureAccepted()` — the `MessageHandler` conversion is meaningful again, and the queue keeps draining while a limited transport is over budget.

Also in this push: the retry delay is clamped with `max(0, …)` (limiters floor `retryAfter` to the second, so the diff can hit 0 right before a reset); two behavior tests, since the only test on this path asserted wiring — the transport throws once the limit is hit, and the handler rethrows as recoverable with the delay; a leftover 7.4 changelog entry from the rebase is removed; wording in the config `info()` and the `setRateLimiterFactory()` docblock now describes the fail-fast behavior.

One note for the docs PR: `RateLimitExceededException` is not a `TransportExceptionInterface`, so `FailoverTransport` won't fail over on a rate limit — spilling to the next transport could be a later opt-in feature.

Mailer is green at 958 tests, FrameworkBundle DI at 583.

nicolas-grekas commented:
Thank you @fritzmg.
